enable irqsteer for m33 of imx943

biwenli opened the pull request:
enable irqsteer for imx943.

iuliana-prodan commented on drivers/interrupt_controller/CMakeLists.txt:
```diff
@@ -44,6 +44,7 @@ zephyr_library_sources_ifdef(CONFIG_NXP_PINT                intc_nxp_pint.c)
 zephyr_library_sources_ifdef(CONFIG_RENESAS_RX_ICU          intc_renesas_rx_icu.c)
 zephyr_library_sources_ifdef(CONFIG_RENESAS_RZ_EXT_IRQ      intc_renesas_rz_ext_irq.c)
 zephyr_library_sources_ifdef(CONFIG_NXP_IRQSTEER            intc_nxp_irqsteer.c)
+zephyr_library_sources_ifdef(CONFIG_NXP_IRQSTEER1           intc_nxp_irqsteer1.c)
```
Please explain why do you need a new driver? 
Why the `intc_nxp_irqsteer.c` driver is not enough?

In the commit message you state that the "old" `intc_nxp_irqsteer` driver can be replaced.
In this case, why don't you enhance the `intc_nxp_irqsteer` driver with the support that you need for i.MX943?

biwenli replied:
> @biwenli so your proposal is to replace the "old" irqsteer driver with the new one?
> 
> Usually, when such a thing is happening you need to update all the references to the old driver with the new one. And as last point remove the old driver.

Yes, the old driver will be phased out over time. New product will use the new driver directly.

iuliana-prodan replied:
@biwenli Could you please add a note indicating that the old `irqsteer` will be deprecated and should no longer be used in new samples or products?
Also, it would be helpful to include at least one commit where the old `irqsteer` is replaced with the new implementation, so others can refer to it as an example.
Thanks.

biwenli replied:
> @biwenli Could you please add a note indicating that the old `irqsteer` will be deprecated and should no longer be used in new samples or products? Also, it would be helpful to include at least one commit where the old `irqsteer` is replaced with the new implementation, so others can refer to it as an example. Thanks.

@iuliana-prodan ,

Sure, np. Thanks.

dbaluta commented:
@biwenli you will also need to convert at least 1 user of the old driver to test your new implemenation. Ideally would be to convert all users, but we can do it in phases as we don't want to slow you down.

biwenli committed:
boards: nxp: imx943_evk: m33: enable edma4

Enable edma4 by default for m33(in NETCMIX) of imx943_evk

Signed-off-by: Biwen Li <biwen.li@nxp.com>

mmahadevan108 commented on modules/hal_nxp/mcux/mcux-sdk-ng/drivers/drivers.cmake:
```diff
@@ -142,9 +142,14 @@ set_variable_ifdef(CONFIG_MIPI_DBI_NXP_DCNANO_LCDIF CONFIG_MCUX_COMPONENT_driver
+if(NOT CONFIG_SOC_MIMX9596)
+  set_variable_ifdef(CONFIG_ETH_NXP_IMX_NETC          CONFIG_MCUX_COMPONENT_driver.netc_switch)
+endif()
+
```
Please fix this commit. Either drop/squash with the earlier change/fix the commit message.

biwenli replied:
> Please fix this commit. Either drop/squash with the earlier change/fix the commit message.

fixed.

biwenli committed:
irqsteer-master: Introduce priority attribute

This adds priority attribute to irqsteer in order to
support multi level interrupts.

We need to make this change atomic so that we don't break
the build.

So, we introduce a new property named 'priority' to
`interrupt-cells` and we increase the number of cells.
Then we update all instances of interrupts
using irqsteer (for imx95, imx8m, imx8qm, imx8qxp and also
update the overlay(imx8mp_evk_mimx8ml8_adsp) for tests.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
samples: drivers: uart_aysnc_api: imx943_evk: m33: add config

Add config for imx943_evk_mimx94398_m33
- Use the sample to test the irqsteer driver

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli commented:
> > @biwenli you will also need to convert at least 1 user of the old driver to test your new implemenation. Ideally would be to convert all users, but we can do it in phases as we don't want to slow you down.
> 
> Sure, np.



> @biwenli you will also need to convert at least 1 user of the old driver to test your new implemenation. Ideally would be to convert all users, but we can do it in phases as we don't want to slow you down.

biwenli commented:
> > > @biwenli you will also need to convert at least 1 user of the old driver to test your new implemenation. Ideally would be to convert all users, but we can do it in phases as we don't want to slow you down.
> > 
> > 
> > Sure, np.
> 
> > @biwenli you will also need to convert at least 1 user of the old driver to test your new implemenation. Ideally would be to convert all users, but we can do it in phases as we don't want to slow you down.

The test case can be used to test the new irqsteer1 driver,
west build -p always -b imx943_evk/mimx94398/m33 samples/drivers/uart/async_api -d ./build_samples_uart_async_api

dbaluta commented on tests/drivers/build_all/interrupt_controller/common/boards/imx8mp_evk_mimx8ml8_adsp.overlay:
```diff
@@ -22,23 +22,23 @@
 			    compatible = "nxp,irqsteer-master";
 			    reg = <0>;
 			    interrupt-controller;
```
Please provide more context in your commit messages. One year later a poor guy who tries to understand this patch wont find any useful info.

biwenli committed:
tests: app_dev: ram_context_for_isr: fix irq_num

This commit fixes build issue when multi level interrupts is enabled,
define TEST_IRQ_NUM as (CONFIG_2ND_LVL_ISR_TBL_OFFSET - 1)
  - CMSIS/Core/Include/core_cm33.h:2438:15: error: array subscript 24
    is above array bounds of volatile uint32_t[16]

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
tests: arch: arm: irq_advanced_features: fix build issue

This commit fixes build issue when multi level interrupts feature
is enabled, define new macro TEST_1ST_LEVEL_INTERRUPTS_MAX to
support multi level interrupts case.
  - error: array subscript 24 is above array bounds of 'volatile
    uint32_t[16]' {aka 'volatile unsigned int[16]'}
    [-Werror=array-bounds])

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
tests: arch: arm: custom_interrupt: fix build issue

Fix build issue when multi level interrupts is enabled,
define new macro TEST_1ST_LEVEL_INTERRUPTS_MAX to support multi
level interrupts case.
  - core_cm33.h:2559:47: error: iteration 496 invokes undefined behavior
    [-Werror=aggressive-loop-optimizations]

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
tests: arch: arm_custom_interrupt: exclude imx943

Exclude imx943,
  - multi level interrupts feature is enabled defaultly,
    so irqsteer1 driver is enabled, then the test case
    report the build issue:
    multiple definition of z_soc_irq_enable,
    tests/arch/arm/arm_custom_interrupt/src/arm_custom_interrupt.c:47
    first defined here.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

decsny reviewed (dismissed):
**driver looks like almost an exact copy paste of the other one**. figure out how to reuse the code. this is not okay to be copy pasting drivers and ballooning the amount of code to maintain and duplicating the maintenance effort requirement across every chip.

dbaluta commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -546,7 +546,7 @@
 				compatible = "nxp,irqsteer-master";
 				reg = <0>;
 				interrupt-controller;
-				#interrupt-cells = <1>;
```
There should be no build issue because you break bisectability. You should put this in the same commit that updates the interrupt-cells and mention this. 

(We are adding all the code here to avoid breaking the build). Remember, a patch should do one logical change per patch.

biwenli commented:
Hi Guys,

Any comments?

decsny commented on drivers/interrupt_controller/Kconfig.nxp_irqsteer:
```diff
@@ -11,3 +11,8 @@ config NXP_IRQSTEER
 		more CPU interrupt lines. This is used for CPUs
 		such as XTENSA DSPs.
+
+config USE_NXP_MCUX_SDK_IRQSTEER1
+	bool "Use IRQSTEER1 driver in NXP MCUX SDK"
+	default n
+	depends on NXP_IRQSTEER
```
is this a different version of this IRQSTEER or a completely different IP? because for example with EDMA we use different drivers but put a version property in DT to pick which one, take a look at EDMA kconfigs file for example

biwenli replied:
> is this a different version of this IRQSTEER or a completely different IP? because for example with EDMA we use different drivers but put a version property in DT to pick which one, take a look at EDMA kconfigs file for example

Fixed it after following the rules from mcux edma driver.

dleach02 commented:
PR posted to fix CI failure (#96308)

JiafeiPan commented:
Hi, @biwenli , please fix twister failures.

dleach02 commented:
@biwenli I rebased your PR and pushed it to see if the CI failures clear

biwenli committed:
drivers: intc: irqsteer: support irqsteer v1 in mcux sdk

There are two irqsteer driver in mcux sdk,
Their apis are not compatible. So add new option
CONFIG_INTC_MCUX_IRQSTEER_V1
to pick right irqsteer driver in mcux sdk for
the zephyr irqsteer driver

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
dts: arm: nxp: imx943: m33: add irqsteer node

Add irqsteer node to support
multi level interrupts

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
dts: bindings: intc: irqsteer: add new property nxp,version

Add new property nxp,version to
pick right irqsteer driver in nxp mcux sdk
(There are two irqsteer driver in nxp mcux sdk,
and their apis are not compatible).

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
modules: hal_nxp: mcux: mcux-sdk-ng: pick right irqsteer driver

Pick right irqsteer driver in nxp mcux sdk.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
soc: nxp: imx943: m33: enable multi level interrupts

This commit enables multi level interrupts:
- Enable multi level interrupts
  imx943 is a two level interrupts system and
  some interrupts(eg. edma interrupts) depend on irqsteer.
- Increase irq number from 405 to 790
  Actually there are 407(IRQ 0 ~ IRQ 406) interrupts
  from nvic(first level interrupts),
  The second level interrupts are extended by irqsteer,
  it extends 32 x 12 = 384 interrupts,
  So first level interrupts + second level interrupts
  = 407 + 384 = 791(IRQ 0 ~ IRQ 790) interrupts

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -546,7 +546,7 @@
 				compatible = "nxp,irqsteer-master";
```
> I don't understand this change sorry! You did merge the commits but you didn't updated the commit message! You just merged all the previous commit messages including the format with 5 commit title like messages.
> 
> Please write a proper commit message as this would be a single change. Explain why the change was needed.
> 
> Please don say "update interrupt-cells" say something like "Increase the number of interrupt-cells because here.

sure, np.

biwenli commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -871,6 +871,8 @@ static int dma_mcux_edma_get_status(const struct device *dev, uint32_t channel,
 	LOG_DBG("DMA CHx_INT 0x%x", DEV_BASE(dev)->TCD[hw_channel].CH_INT);
 	LOG_DBG("DMA TCD_CSR 0x%x", DEV_BASE(dev)->TCD[hw_channel].CSR);
+	LOG_DBG("DMA TCD_SADDR 0x%x", DEV_BASE(dev)->TCD[hw_channel].SADDR);
+	LOG_DBG("DMA TCD_DADDR 0x%x", DEV_BASE(dev)->TCD[hw_channel].DADDR);
```
> This is a trivial fix and it really doesn't belong to this PR as this PR is totally about a different thing. Yo ucan send this as a separate PR mark it as trivial and have it merge immediately.

sure, np.

dbaluta commented on soc/nxp/imx/imx9/imx943/Kconfig.defconfig.mimx94398.m33:
```diff
@@ -14,8 +14,52 @@ config FLASH_BASE_ADDRESS
 config MCUX_CORE_SUFFIX
 	default "_cm33_core1" if SOC_MIMX94398_M33
 
+# multi-level interrupts
```
This is OK now, it is nicely explaining why do we need the change.

biwenli commented on dts/arm/nxp/nxp_imx95_m7.dtsi:
```diff
@@ -546,7 +546,7 @@
 				compatible = "nxp,irqsteer-master";
 				reg = <0>;
```
> Please squash this commit into a single unit. Here the commit message looks like is the merge of several other commits.
> 
> Commit should look like this:
> 
> ```
> 
> irqsteer-master: Introduce priority attribute
> 
> This adds priority attribute to irqsteer in order to support multi level interrupts.
> 
> We need to make this change atomic so that we don't break the build.
> 
> So, we introduce a new property named 'priority' to `interrupt-cells` 
> and we increase the number of cells. Then we update all instances 
> of interrupts using irqsteer (for imx95, imx8m, imx8, imx8) and also 
> update the overlay for tests.
> ```

sure, np.

biwenli committed:
dts: arm: nxp: imx943: m7_0: add irqsteer node

Add irqsteer node for m7_0.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

biwenli committed:
dts: arm: nxp: imx943: m7_1: add irqsteer node

Add irqsteer node for m7_1.

Signed-off-by: Biwen Li <biwen.li@nxp.com>

kylebonnici commented:
@biwenli  PR is merged rebase and try again

LaurentiuM1234 reviewed (changes requested):
NACK on the current version. Reasons listed below:

1. Commits not sequenced properly.
2. Commits break bisectability.
3. DT properties should be used to describe hardware properties, not distinguish between driver revisions.

Also, IMO, this highlights the following problem: the HAL API changes are done with no regard for the upper layers. We can't have to constantly modify the upper layers because some new HAL API change breaks compatibility with the old API.
